### PR TITLE
Search for bootcamps

### LINKED ISSUE #2043: Update frontend code to additionally search for `bootcamp` record types
Support the new LearningResource types on the frontend
- [ ] Add facets for the `LearningResource` types
- [ ] Allow filtering on courses
- [ ] allow filtering on bootcamps

### PR DESCRIPTION
**Pull request by mbertrand**
#### Pre-Flight checklist

- [x] Screenshots and design review for any changes that affect layout or styling
  - [x] Desktop screenshots
  - [x] Mobile width screenshots
  - [x] Tag @ferdi or @pdpinch for review
- [x] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested

#### What are the relevant tickets?
Fixes #2043 

#### What's this PR do?
- Modifies the frontend search query code to handle bootcamps and courses.
- Adds a search facet for 'learning resource' that includes courses and bootcamps
- Adds CompactBootcampDisplay and ExpandedBootcampDisplay components to display Bootcamp results
- renames `CourseDrawer` to `LearningResourceDrawer` and modifies it to handle different object types instead of just courses.

#### How should this be manually tested?
- Run `docker-compose run web python manage.py recreate_index`
- Import some bootcamps if you haven't already done so:
```python
from course_catalog.tasks import *
get_bootcamp_data()
```
- Import some course data if you haven't already done so:
```
sync_and_upload_edx_data()
```
- Go to `/courses/search`.  There should be a new facet for `Learning Resource` that includes checkboxes for 'Course' and 'Bootcamp'.  Try these out and ensure that you can see both Course and Bootcamp records returned, and that clicking on a result opens the right-side drawer containing the expanded display for the course/bootcamp.

#### Screenshots (if appropriate)
**Learning Resource Facet**
Screen Shot 2019-07-23 at 2 18 53 PM

**Learning Resource Facet, mobile**
Screen Shot 2019-07-23 at 2 19 35 PM

**Bootcamp facet selected, bootcamp card displayed**
Screen Shot 2019-07-23 at 2 19 06 PM

** Bootcamp expanded display **
Screen Shot 2019-07-23 at 2 32 20 PM

### CODE REVIEW

**Comment by mbertrand** [comment]
@abdulkdawson  @Ferdi @pdpinch screenshots added

File: static/js/components/CourseCard.js [hunk 1]
```diff
@@ -41,7 +42,7 @@ const CourseCard = ({
             alt={`cover image for ${course.title}`}
           />
         </div>
-        <div className="row course-title" onClick={showCourseDrawer}>
+        <div className="row course-title" onClick={showResourceDrawer}>
```
Comment (aliceriot) [inline static/js/components/CourseCard.js]: Is there some reason we need a separate card type for bootcamps? I think if we can do it it would be better to rename this `CourseCard` into `LearningResourceCard` and then modify it to support both courses and bootcamps, since we're displaying basically the same stuff for them. What do you think?

File: static/js/components/ExpandedBootcampDisplay.js [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+const BOOTCAMP_IMAGE_DISPLAY_HEIGHT = 239
+const BOOTCAMP_IMAGE_DISPLAY_WIDTH = 440
+const entities = new AllHtmlEntities()
+
+type Props = {
+  bootcamp: Bootcamp
+}
+
+export default class ExpandedBootcampDisplay extends React.Component<Props> {
```
Comment (aliceriot) [inline static/js/components/ExpandedBootcampDisplay.js]: I have the same question here - if it's possible to have an `ExpandedLearningResourceDisplay` instead of two different very similar components I think that would be preferable

File: static/js/components/CourseCard.js (hunk 1 above)
Comment (mbertrand) [reply static/js/components/CourseCard.js]: They are very similar but some of the info is different - for example bootcamps do not have a `semester` or `platform`.  I initially considered adding some conditional statements for these to `CourseCard` but ultimately decided to create separate card components in case they ever diverged further, and also because the `objectType` parameter sent to `LearningResourceDrawer` from the card needs to be different.  But if you think it's better to do the conditional approach I can change it.

Comment (aliceriot) [reply static/js/components/CourseCard.js]: I think if they need to diverge further in the future we can create separate components then, but unless we need to do that it's better to add a few small conditionals and avoid writing extra code.

File: static/js/components/CourseCard.js [hunk 3]
```diff
@@ -63,16 +65,20 @@ const CourseCard = ({
...
+          <div className="row platform">
+            <img
+              className="course-platform"
+              // $FlowFixMe: only courses will end up here
+              src={platformLogoUrls[object.platform]}
+              // $FlowFixMe: only courses will end up here
+              alt={`logo for ${object.platform}`}
+            />
+          </div>
+        ) : null}
```
Comment (aliceriot) [inline static/js/components/CourseCard.js]: fyi in the designs we do want to have the logo show up for bootcamp items here

Comment (mbertrand) [reply static/js/components/CourseCard.js]: @abdulkdawson can you send me the logo image for bootcamps?

Comment (aliceriot) [reply static/js/components/CourseCard.js]: sorry - should have clarified, it's already added, if you look at the `platformLogoUrls` object in `constants.js` you can find the URL for it there

Comment (mbertrand) [reply static/js/components/CourseCard.js]: thanks, added the logo for bootcamps

File: static/js/components/ExpandedBootcampDisplay.js (hunk 2 above)
Comment (mbertrand) [reply static/js/components/ExpandedBootcampDisplay.js]: ✔️

File: static/js/components/SearchResult.js [hunk 4]
```diff
@@ -83,22 +89,42 @@ const ProfileSearchResult = ({ result }: ProfileProps) => {
...
+  )
+}
+
+const BootcampSearchResult = ({
+  result,
+  toggleFacet,
+  setShowResourceDrawer
+}: LearningResourceProps) => {
+  // $FlowFixMe: this should only be used for bootcamps
+  const bootcamp = searchResultToBootcamp(result)
```
Comment (aliceriot) [inline static/js/components/SearchResult.js]: one other approach I was thinking that we could take here is to, instead of having a separate `searchResultToBootcamp` we could modify the existing `searchResultToCourse` to be `searchResultToLearningResource` to sort of follow-along with how the UI is organized. That function could set `platform` to be `bootcamps` when the `object_type` is `bootcamp`, and then that value could be used throughout the UI to control whether we show bootcamp-specific UI or not.

Does that make sense? if we did it that way we'd avoid the need for a separate "bootcamp" search result, we wouldn't need to pass down the "objectType" parameter, etc because anywhere we need to figure out how to render the thing we could just look at the `platform` value.

File: static/js/containers/CourseIndexPage.js [hunk 5]
```diff
@@ -123,7 +122,7 @@ const mapStateToProps = (state: Object): StateProps => ({
 })
 
 const mapDispatchToProps = {
-  setShowCourseDrawer
+  setShowResourceDrawer: setShowResourceDrawer
```
Comment (aliceriot) [inline static/js/containers/CourseIndexPage.js]: if you want to you can leave out the colon here, it's called a 'shorthand property name' https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer#New_notations_in_ECMAScript_2015

File: static/js/components/CourseCard_test.js [hunk 6]
```diff
@@ -1,49 +1,56 @@
...
-    })
+    renderCourseCard = ({ ...props }) =>
+      shallow(
+        <CourseCard
+          object={course}
+          objectType="course"
+          setShowResourceDrawer={setShowResourceDrawerStub}
+          {...props}
+        />
+      )
```
Comment (aliceriot) [inline static/js/components/CourseCard_test.js]: any reason why you refactored away from the `configureShallowRenderer` here?

File: static/js/factories/resources.js [hunk 7]
```diff
@@ -0,0 +1,76 @@
+//@flow
```
Comment (aliceriot) [inline static/js/factories/resources.js]: might be good to call this file `learning_resources` so it's more immediately clear what it's about

File: static/js/lib/search.js [hunk 8]
```diff
@@ -192,7 +227,11 @@ export const buildSearchQuery = ({
     builder.sort(field, option)
   }
 
-  const types = type ? [type] : ["comment", "post", "profile"]
+  const types = type
+    ? Array.isArray(type)
+      ? type
+      : [type]
+    : ["comment", "post", "profile"]
```
Comment (aliceriot) [inline static/js/lib/search.js]: if possible it might be good to refactor away from a nested ternary, just a bit hard to follow

File: static/js/lib/util.js [hunk 9]
```diff
@@ -148,3 +148,11 @@ export const flatZip = R.compose(
...
   R.zip
 )
+
+export const capitalize = R.converge(R.concat(), [
+  R.compose(
+    R.toUpper,
+    R.head
+  ),
+  R.tail
+])
```
Comment (aliceriot) [inline static/js/lib/util.js]: what does this function do differently than `_.capitalize`? I tried it out in a node repl and `_.capitalize` seems to return the same thing

File: static/js/reducers/bootcamps.js [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+  return update
+}
+
+export const bootcampsEndpoint = {
+  name:              "bootcamps",
+  verbs:             [GET],
+  initialState:      { ...INITIAL_STATE, data: new Map() },
+  getFunc:           (id: number) => bootcampAPI.getBootcamp(id),
+  getSuccessHandler: updateBootcampHandler
+}
```
Comment (aliceriot) [inline static/js/reducers/bootcamps.js]: I think it's not necessary for this PR, but if you want to use redux-query for managing this API instead of redux-hammock I think that would be a good idea, we're generally migrating away from redux-hammock and towards redux-query. I've been using it for any new APIs in Open for a little bit.

Comment (aliceriot) [reply static/js/reducers/bootcamps.js]: If you don't want to take the time now though that's alright, this is a pretty simple API so we can easily change it out later.

File: static/js/lib/util.js (hunk 9 above)
Comment (mbertrand) [reply static/js/lib/util.js]: In an earlier PR review, Nathan said I should avoid using lodash for new functionality and use ramda instead, so this is the ramda equivalent.

File: static/js/components/SearchResult.js (hunk 4 above)
Comment (mbertrand) [reply static/js/components/SearchResult.js]: I can make a `searchResultToLearningResource` that then calls the existing `searchResultToCourse` or `searchResultToBootcamp` function depending on the `object_type`.  The `Bootcamp` type does not include `platform` or `level` or `semester`), and not sure if `platform` makes sense anyway for the upcoming `LearningPath` results, which will be very different than bootcamps and courses.

File: static/js/lib/util.js (hunk 9 above)
Comment (aliceriot) [reply static/js/lib/util.js]: ah ok, I was just wondering because I saw some other uses of `_.capitalize` around as well.

File: static/js/components/SearchResult.js (hunk 4 above)
Comment (aliceriot) [reply static/js/components/SearchResult.js]: yeah, I was thinking instead of even having separate `Course` and `Bootcamp` types why not have a single `LearningResource`? the two `Course` and `Bootcamp` types you have now have the majority of their fields in common already.

Comment (mbertrand) [reply static/js/components/SearchResult.js]: Don't I need different types since they have different fields, even if currently it's just a few, otherwise flow will start complaining? I could have a `LearningResource` that both the `Course` and `Bootcamp` types inherit from with common fields, but once I introduce `Program` and `LearningPath` the common fields will go down substantially.

Comment (mbertrand) [reply static/js/components/SearchResult.js]: Maybe I'm misunderstanding but I'm trying to emulate the current code pattern where for example there are separate types for `Post` and `Comment` (though they have `AuthoredContent` type for common fields) and separate `searchResultToPost` and `searchResultToComment` functions to convert the ES search results to their respective types.

Comment (aliceriot) [reply static/js/components/SearchResult.js]: to get around flow complaining fields can be typed as optional. yeah, I'm not sure exactly how we do it, I'm just seeing how much code is being added just to add one more type of result (bootcamps) and that has me worried about dealing with tons of similar code when we add programs and learning paths as well, so I think if we can get out in front of it by trying to re-use as much as possible we'll be in a better position going forward.

it seems to me that the display we want to have in the learning resource card is mostly the same for all of these types of resources, and then there are differences in how we want to display them in the learning resource drawer. to me the way to handle that that makes things easiest is to treat these things the same way everywhere except where we actually need to render something different, like the list of things in the drawer display for a learning path or similar. it's a bit analogous to how we don't worry about whether a post is a link post or a text post unless we're actually at the point of rendering it's content.

anyhow, if you don't agree that's fine, we can keep this as-is for now. I just want to make sure we don't end up with a lot of very similar code for handling these different things that's hard to maintain later.

Comment (mbertrand) [reply static/js/components/SearchResult.js]: I think I might understand now, the `LearningResource` type would be independent of the existing `Bootcamp` and `Course` types (still needed for the expanded display in the drawer) and have only the fields displayed in the compact card?

File: static/js/lib/search.js (hunk 8 above)
Comment (mbertrand) [reply static/js/lib/search.js]: ✔️

File: static/js/reducers/bootcamps.js (hunk 10 above)
Comment (mbertrand) [reply static/js/reducers/bootcamps.js]: I think I'd rather work on that for the next PR when I introduce the API's for learning paths and programs.  Will probably need to add `reselect` for redux-query to package.json & yarn.lock, plus rewrite the course API code, and this PR is already pretty big.

File: static/js/components/SearchResult.js (hunk 4 above)
Comment (mbertrand) [reply static/js/components/SearchResult.js]: I did the above

File: static/js/reducers/bootcamps.js (hunk 10 above)
Comment (aliceriot) [reply static/js/reducers/bootcamps.js]: sounds good, I'm adding reselect and fiddling a bunch with the courses API currently anyhow in my work for #2091
